plugins/providers/virtualbox/cap/cleanup_disks.rb

```diff
@@ -25,26 +26,77 @@ def self.cleanup_disks(machine, defined_disks, disk_meta_file)
         # @param [VagrantPlugins::Kernel_V2::VagrantConfigDisk] defined_disks
         # @param [Hash] disk_meta - A hash of all the previously defined disks from the last configure_disk action
         def self.handle_cleanup_disk(machine, defined_disks, disk_meta)
-          vm_info = machine.provider.driver.show_vm_info
-          primary_disk = vm_info["SATA Controller-ImageUUID-0-0"]
-
-          disk_meta.each do |d|
-            dsk = defined_disks.select { |dk| dk.name == d["name"] }
-            if !dsk.empty? || d["uuid"] == primary_disk
-              next
-            else
-              LOGGER.warn("Found disk not in Vagrantfile config: '#{d["name"]}'. Removing disk from guest #{machine.name}")
-              disk_info = machine.provider.driver.get_port_and_device(d["uuid"])
+          storage_controllers = machine.provider.driver.read_storage_controllers
+          if storage_controllers.size ==  1
+            primary_controller = storage_controllers.first
+          else
+            primary_controller = storage_controllers.detect { |c| c.storage_bus == "SATA" }
+            if primary_controller.nil?
+              raise Vagrant::Errors::VirtualBoxDisksControllerNotFound, storage_bus: "SATA"
+            end
+          end
 
-              machine.ui.warn("Disk '#{d["name"]}' no longer exists in Vagrant config. Removing and closing medium from guest...", prefix: true)
+          primary = primary_controller.attachments.detect { |a| a[:port] == "0" && a[:device] == "0" }
+          if primary.nil?
+            raise Vagrant::Errors::VirtualBoxDisksPrimaryNotFound
+          end
+          primary_uuid = primary[:uuid]
 
-              if disk_info.empty?
-                LOGGER.warn("Disk '#{d["name"]}' not attached to guest, but still exists.")
+          if disk_meta
```

Comment:
Hm, I don't think this if is required. It should be an empty hash if there's no meta data, so the each will exit without enumerating over anything already.